tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,305 @@
+    registry::{LookupMetadata, LookupSpan, SpanRef},
+};
+use std::{cell::RefCell, fmt, io, marker::PhantomData};
+use tracing_core::{
+    span::{Attributes, Id, Record},
+    Event, Subscriber,
+};
+
+/// A [`Layer`] that logs formatted representations of `tracing` events.
+///
+/// [`Layer`]: ../trait.Layer.html
+#[derive(Debug)]
+pub struct FmtLayer<
+    S,
+    N = format::DefaultFields,
+    E = format::Format<format::Full>,
+    W = fn() -> io::Stdout,
+> {
+    make_writer: W,
+    fmt_fields: N,
+    fmt_event: E,
+    _inner: PhantomData<S>,
+}
+
+/// A builder for `FmtLayer` that logs formatted representations of `tracing`
+/// events.
+#[derive(Debug)]
+pub struct Builder<
+    S,
+    N = format::DefaultFields,
+    E = format::Format<format::Full>,
+    W = fn() -> io::Stdout,
+> {
+    fmt_fields: N,
+    fmt_event: E,
+    make_writer: W,
+    _inner: PhantomData<S>,
+}
+
+impl<S> FmtLayer<S> {
+    /// Creates a [Builder].
+    pub fn builder() -> Builder<S> {
+        Builder::default()
+    }
+}
+
+// This, like the MakeWriter block, needs to be a seperate impl block because we're
+// overriding the `E` type parameter with `E2`.
+impl<S, N, E, W> Builder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    W: MakeWriter + 'static,
+{
+    /// Sets a [FormatEvent<S, N>].
+    pub fn with_event_formatter<E2>(self, e: E2) -> Builder<S, N, E2, W>
+    where
+        E2: FormatEvent<S, N> + 'static,
+    {
+        Builder {
+            fmt_fields: self.fmt_fields,
+            fmt_event: e,
+            make_writer: self.make_writer,
+            _inner: self._inner,
+        }
+    }
+}
+
+// this needs to be a seperate impl block because we're re-assigning the the W2 (make_writer)
+// type paramater from the default.
+impl<S, N, E, W> Builder<S, N, E, W> {
+    /// Sets a [MakeWriter] for spans and events.
+    pub fn with_writer<W2>(self, make_writer: W2) -> Builder<S, N, E, W2>
+    where
+        W2: MakeWriter + 'static,
+    {
+        Builder {
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            make_writer,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl<S, N, E, W> Builder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    /// Builds a [FmtLayer] infalliably.
+    pub fn build(self) -> FmtLayer<S, N, E, W> {
+        FmtLayer {
+            make_writer: self.make_writer,
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl<S> Default for FmtLayer<S>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+{
+    fn default() -> Self {
+        Builder::default().build()
+    }
+}
+
+impl<S> Default for Builder<S> {
```

Comment:
it's weird how impls on the layer and on the builder are interleaved here, can we sort them by type?